src/jit/decomposelongs.cpp

```diff
@@ -401,25 +404,40 @@ GenTree* DecomposeLongs::DecomposeStoreLclVar(LIR::Use& use)
 
     GenTree* loRhs   = rhs->gtGetOp1();
     GenTree* hiRhs   = rhs->gtGetOp2();
+
+    // Keep the original store's flags except side effects, the store
+    // itself should not have any. The lo and hi rhs parts could, at least
+    // in theory, have different side effects so we can't simply copy the
+    // original store's side effecs to both parts.
+    unsigned flags   = tree->gtFlags & ~GTF_ALL_EFFECT;
+    unsigned loFlags = flags | (loRhs->gtFlags & GTF_ALL_EFFECT);
+    unsigned hiFlags = flags | (hiRhs->gtFlags & GTF_ALL_EFFECT);
```

Comment:
The store nodes should also have the `GTF_ASG` flag set. We should probably validate that all stores have this flag set in `CheckLIR`, but that change should be made separately.